### PR TITLE
Port to GNOME Software 42

### LINKED ISSUE #17: Document changes for GNOME Software 42
This is just a documenting issue to keep track of the changes to the plugin structure in the upcoming GNOME Software 42. They are doing a lot of work and it looks like many things will change. I am just listing here the MRs that I found relevant so that we can have a roadmap for future upgrade:

- [ ] Plugins are now subclasses: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1042
 - [ ]  Rework get and set management plugin: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1106
 - [ ] Async plugin setup and shutdown: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1114
 - [ ] Changed refine flags DEFAULT and USE_HISTORY: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1132
 - [ ] Rework of refine to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1134
 - [ ] Rework of add_installed to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1169
 - [ ] Change cache to `guint64` and rename for better readability: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1219
 - [ ] Rework of refresh to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1259

If anybody discovers anything else that changed, comments are more than welcomed.

### PR DESCRIPTION
**Pull request by DylanVanAssche**
Ports the plugin to the new API in GNOME Software 42:

- [x] Plugins are now subclasses: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1042
- [x] Rework get and set management plugin: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1106
- [x] Async plugin setup and shutdown: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1114
- [x] Changed refine flags DEFAULT and USE_HISTORY: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1132
- [x] Rework of refine to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1134
- [x] Rework of add_installed to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1169
- [x] Change cache to guint64 and rename for better readability: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1219
- [x] Rework of refresh to make it asynchronous: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1259

See individual commits for a detailed description of the changes.
Not tested at all, will comment here when done.

Fixes #17

### CODE REVIEW

**Comment by DylanVanAssche** [comment]
Did a smoke test and everything works fine from the first try :tada:

**Comment by DylanVanAssche** [comment]
Some actions like on startup are still synchronous actions, blocking the UI thread.
I wouldn't fix it here, but rather in a next MR since this one is already a big change.

**Comment by pabloyoyoista** [comment]
Thank you for taking care of this! The diff is even smaller than I expected. 

> Some actions like on startup are still synchronous actions, blocking the UI thread.
I cannot assess the asynchronous code, I have no idea of how it works, would you mind going a bit in more detail about this? :)

Some additional comments:
 * Do I understand by your comment in IRC/matrix that tests are working fine in your local environment?
 * For CI to run successfully, we still need some fixes to the GH Actions for 42. I'll give myself this week as a deadline to make it work, hope that's fine.
 * Once we get #35 in, there will be some merge conflicts, so we'll need a rebase and re-review once that is in. If we don't manage to do it this week, and to avoid blocking your work, I could push a `42` branch for this to be merged. Then we can unblock your work and rebase/merge on top of master when all the other details are solved.

File: src/gs-plugin-apk/gs-plugin-apk.c [hunk 1]
```diff
@@ -942,3 +946,14 @@ gs_plugin_remove_repo (GsPlugin *plugin,
   gs_app_set_state (repo, GS_APP_STATE_AVAILABLE);
   return TRUE;
 }
+
+static void
+gs_plugin_apk_class_init (GsPluginApkClass *klass)
+{
+}
```
[the remark this answers is not in the document]
Comment (DylanVanAssche) [reply src/gs-plugin-apk/gs-plugin-apk.c:993]: Ah yes! We probably need one which frees all private data. We could do a follow up here.

**Comment by DylanVanAssche** [comment]
> I cannot assess the asynchronous code, I have no idea of how it works, would you mind going a bit in more detail about this? :)

The code is not fully asynchronous yet because the calls to apk-polkit-rs are still synchronous.
So what happens currently:

- Create a GTask to start an async operation `g_task_new`. The provided callback will be called by GTask when the task is complete so you can get the result of the operation (not used yet here!)
- Perform the operation, this is currently a sync operation which needs to be fixed in a follow up.
- Return an error with `g_task_return_error` if needed
- Return the result of the task such as a boolean with `g_task_return_boolean`

Once we make the apk-polkit-rs calls async, we pass the GTask to the callback where we return the result of the task, then the code becomes non blocking & async.

Besides that, there's like no docs on GTask on how to use it I'm afraid, I learned it through ModemManager MRs :)

File: src/gs-plugin-apk/gs-plugin-apk.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (DylanVanAssche) [reply src/gs-plugin-apk/gs-plugin-apk.c:993]: https://github.com/Cogitri/gnome-software-plugin-apk/issues/43

File: tests/gs-self-test.c [hunk 2]
```diff
@@ -182,7 +184,6 @@ gs_plugins_apk_app_install_remove (GsPluginLoader *plugin_loader)
 
   // make sure we got the correct app and is managed by us
   g_assert_cmpstr (gs_app_get_id (app), ==, "apk-test-app.desktop");
-  g_assert_cmpstr (gs_app_get_management_plugin (app), ==, "apk");
```
Comment (pabloyoyoista) [inline tests/gs-self-test.c:185]: Here we should also still check that the management plugin is the correct one :)

File: src/gs-plugin-apk/gs-plugin-apk.c [hunk 3]
```diff
@@ -528,7 +528,7 @@ set_app_metadata (GsPlugin *plugin, GsApp *app, ApkdPackage *package, GsPluginRe
     {
       gs_app_set_license (app, GS_APP_QUALITY_UNKNOWN, package->m_license);
     }
-  if (flags & GS_PLUGIN_REFINE_FLAGS_DEFAULT)
+  if (flags & GS_PLUGIN_REFINE_FLAGS_NONE)
```
[the remark this answers is not in the document]
Comment (pabloyoyoista) [reply src/gs-plugin-apk/gs-plugin-apk.c:563]: Sorry if I expressed myself wrong. Please remove this if completely and see the comment below

File: src/gs-plugin-apk/gs-plugin-apk.c [hunk 4]
```diff
@@ -685,20 +709,33 @@ resolve_matching_package (GsPlugin *plugin,
...
+                            GsPluginRefineFlags  flags,
+                            GCancellable        *cancellable,
+                            GAsyncReadyCallback  callback,
+                            gpointer             user_data)
+{
+  g_autoptr (GTask) task = NULL;
+  g_autoptr (GError) error = NULL;
+
+  task = g_task_new (plugin, cancellable, callback, user_data);
+  g_task_set_source_tag (task, gs_plugin_apk_refine_async);
```
Comment (pabloyoyoista) [inline src/gs-plugin-apk/gs-plugin-apk.c:733]: Here we need to test against the `NONE` flag. As we should totally skip the refine if that is the only flag set.

```C
if (flags == GS_PLUGIN_REFINE_FLAGS_NONE)
  return TRUE;
```

File: src/gs-plugin-apk/gs-plugin-apk.c [hunk 5]
```diff
@@ -49,7 +49,7 @@ typedef struct
  * g_variant_to_apkd_package:
  * @value_tuple: a GVariant, as received from apk_polkit1_call*
  *
- * Convenience finction which conerts a GVariant we get pack from our DBus
+ * Convenience function which converts a GVariant we get from our DBus
```
[the remark this answers is not in the document]
Comment (pabloyoyoista) [reply src/gs-plugin-apk/gs-plugin-apk.c:57]: Sorry, I meant to not remove the *pack* word, but instead word it like this:
```suggestion
 * Convenience function which converts a GVariant pack we get from our DBus
```

For some reason yesterday I didn't find the suggestion button...

**Review by pabloyoyoista — COMMENTED** [review]
Thanks a lot for the fixes. It's my first time doing a formal review like this one, so sorry if I didn't express things properly at first.

**Comment by pabloyoyoista** [comment]
Thanks a lot for the async explanations, really help :) I had pending reading [this](https://developer.gnome.org/documentation/tutorials/asynchronous-programming.html) if I were to do the porting. Not sure if that is of much help though, as I haven't had the chance to read it...

**Comment by DylanVanAssche** [comment]
> Thanks a lot for the fixes. It's my first time doing a formal review like this one, so sorry if I didn't express things properly at first.

Ah no worries :) You do great! 
I will fix your comments later today.

> Thanks a lot for the async explanations, really help :) I had pending reading [this](https://developer.gnome.org/documentation/tutorials/asynchronous-programming.html) if I were to do the porting. Not sure if that is of much help though, as I haven't had the chance to read it...

Ah that's the first time I found a tutorial, I didn't with the old docs. Finally some good docs on this GTask thingy!

**Comment by pabloyoyoista** [comment]
Also, I'd happily rebase and add the commit for the 42 CI here, but I guess github doesn't have that feature available?

**Comment by DylanVanAssche** [comment]
@pabloyoyoista Maybe we can do it when #35 is merged as well? We need to rebase a lot then.

**Comment by pabloyoyoista** [comment]
Absolutely :)

**Comment by DylanVanAssche** [comment]
Rebased on `master` after #35 was merged.
Still needs to be tested.

**Comment by pabloyoyoista** [comment]
Thank you! I figured out I can push to your branch, so just pushed the CI fix, let's see how it goes.

**Comment by DylanVanAssche** [comment]
@pabloyoyoista Ah okay, but things may fail though :P I rebased and trying to build it locally.

**Comment by pabloyoyoista** [comment]
Great! You can ignore the "Run tests / Test (pull_request)" run. That shouldn't have run. The other one is the one which is meaningful :)

File: src/gs-plugin-apk/gs-plugin-apk.c [hunk 6]
```diff
@@ -663,16 +690,35 @@ refine_apk_package (GsPlugin *plugin,
...
+{
+  g_autoptr (GTask) task = NULL;
+  g_autoptr (GError) error = NULL;
+
+  task = g_task_new (plugin, cancellable, callback, user_data);
+  g_task_set_source_tag (task, gs_plugin_apk_refine_async);
+
+  if (flags == GS_PLUGIN_REFINE_FLAGS_NONE)
+    g_task_return_boolean (task, TRUE);
+  return;
```
Comment (pabloyoyoista) [inline src/gs-plugin-apk/gs-plugin-apk.c:717]: You need braces around this if, otherwise we won't refine :)

Comment (DylanVanAssche) [reply src/gs-plugin-apk/gs-plugin-apk.c]: Done! Can't trigger the CI though because of Github crypto stuff

**Comment by DylanVanAssche** [comment]
Smoke test:

- Searching seems broken? Returns no results (see logs below)
- Getting details about an installed app works
- Searching for updates seems to work, but I don't have any updates currently so I'm not entirely sure, because of this upgrading packages couldn't be tested.

```
18:57:01:0546 Gs  no refine flags set for transaction
18:57:01:0546 Gs  running search on plugin=appstream with dedupe-flags=24 with refine-flags=require-license,require-description,require-rating,require-version,require-history,require-setup-action,require-review-ratings,require-icon,require-permissions with max-results=50 with search=chess, elapsed time since creation 8ms
18:57:01:0547 Gs  no search results to show
```

**Comment by DylanVanAssche** [comment]
Had some pkg upgrades today, upgraded successfully.
So the only thing that seems to be broken is searching for apps.
I think it has something to do with the app refine function, it worked before rebasing on #35 so I think we miss something.
@pabloyoyoista Do you have an idea here, you're more familiar with this part of GS?

**Comment by pabloyoyoista** [comment]
That actually looks more like the appstream data is not properly setup. There is specifically a search call in the tests on purpose so that we could detect this (at the beginning of `gs_plugins_apk_app_install_remove`). In the tests the call looks like this:

```
# Gs-DEBUG: running search on plugin=appstream with dedupe-flags=7 with refine-flags=require-setup-action with timeout=60 with search=apk-test on apps system/package/*/apk-test-app.desktop/*, elapsed time since creation 1ms
```
It sounds strange that it worked before the rebase. The problem that you seem to be having looks related to the appstream plugin not finding apps, not our refine ignoring them (there should otherwise be some debug output from `GsPluginApk`. I guess I'll have to do some testing next week...

**Comment by pabloyoyoista** [comment]
I have done some testing, and seems like the 42 plugin version is doing everything as expected. It took me a while to setup the appstream correctly since again there were some changes. You can maybe check if you get it right building GS from the APKBUILD in [this](https://gitlab.alpinelinux.org/pabloyoyoista/aports/-/tree/gs-42) branch? You would need to force a refresh before getting any data. I am also not sure at this point if it would make more sense to `mv` instead of `rm` the older data though. Once that is done, looking for a chess app

Unfortunately, I have managed to bump into some some segmentation faults related to cancellations of calls (so for example, when fastly typing a search). I don't think this is something to fix in this MR though. So let me know if you can reproduce the successful results, and if you can, we shall get this merged and document any other errors.

**Comment by pabloyoyoista** [comment]
Ok, I have revisited this again and seems like I can basically reproduce the crash whenever there is an action cancellation. Easiest way to reproduce seem to be searching for "gno" or some other short string and waiting for the corresponding timeout. I guess this has be brought upstream, since the crash looks quite deep into GLib's code:

```
#0  magazine_chain_pop_head (magazine_chunks=magazine_chunks@entry=0x7ffff1dcb2f0) at ../glib/gslice.c:583
#1  0x00007ffff7bc2e0e in thread_memory_magazine1_alloc (ix=<optimized out>, tmem=0x7ffff1dcb2e0) at ../glib/gslice.c:886
#2  g_slice_alloc (mem_size=mem_size@entry=16) at ../glib/gslice.c:1063
#3  0x00007ffff7c8d2ca in g_object_notify_queue_freeze (object=object@entry=0x7ffff2e01370 [GsPluginJobRefine], conditional=conditional@entry=0) at ../gobject/gobject.c:294
#4  0x00007ffff7c8e3df in g_object_unref (_object=<optimized out>) at ../gobject/gobject.c:3632
#5  g_object_unref (_object=0x7ffff2e01370) at ../gobject/gobject.c:3553
#6  0x00007ffff7f3b0b9 in gs_plugin_loader_process_thread_cb (task=0x7fffef5d00f0 [GTask], object=<optimized out>, task_data=0x7ffff1a7a140, cancellable=<optimized out>) at ../lib/gs-plugin-loader.c:3366
#7  0x00007ffff7d66ef8 in g_task_thread_pool_thread (thread_data=0x7fffef5d00f0, pool_data=<optimized out>) at ../gio/gtask.c:1441
#8  0x00007ffff7bcc7a3 in g_thread_pool_thread_proxy (data=<optimized out>) at ../glib/gthreadpool.c:354
#9  0x00007ffff7bcc108 in g_thread_proxy (data=0x7ffff2450360) at ../glib/gthread.c:827
#10 0x00007ffff7fba1f5 in start (p=0x7ffff137aca0) at src/thread/pthread_create.c:203
#11 0x00007ffff7fbc470 in __clone () at src/thread/x86_64/clone.s:22
```

@DylanVanAssche have you been able to reproduce? Do you time and energy to take care of this?

**Comment by DylanVanAssche** [comment]
@pabloyoyoista I won't have time to look into for the upcoming time I'm afraid.

**Comment by pabloyoyoista** [comment]
Ok, no problem at all. All the changes here are already super helpful and will help move this forward! I'll start by trying to create a test-case where we execute a cancellable and see if I can reproduce CI. I'll probably bring the question upstream afterwards :) Thank you for you work in this project!

**Comment by pabloyoyoista** [comment]
I had some brief conversation [with Philip](https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1396#note_1475534), seems like priority number 1 should be making the refine async. I'll see if I can get to it next week.

**Comment by pabloyoyoista** [comment]
I have pushed a small fixup commit for a small issue, but otherwise this is great.

@Cogitri I am merging this because having it opened prevents moving forward. The only issue I detected is a crash in searches that I reported in #49 and that we will have to take upstream. In case we want to continue developing GNOME 41, I pushed a gs-41 branch. #49 is certainly a blocker for the release, but hopefully upstream can shed some light onto it :)

**Comment by pabloyoyoista** [comment]
And @DylanVanAssche thanks a lot for taking care of this!
